// inngest/astro.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serve = exports.frameworkName = void 0;
const InngestCommHandler_1 = require("./components/InngestCommHandler");
exports.frameworkName = "astro";
/**
 * In Astro, serve and register any declared functions with Inngest, making
 * them available to be triggered by events.
 *
 * @example
 * ```ts
 * export const { GET, POST, PUT } = serve({
 *   client: inngest,
 *   functions: [fn1, fn2],
 * });
 * ```
 *
 * @public
 */
const serve = (options) => {
    const commHandler = new InngestCommHandler_1.InngestCommHandler(Object.assign(Object.assign({ frameworkName: exports.frameworkName, fetch: fetch.bind(globalThis) }, options), { handler: ({ request }) => {
            return {
                body: () => request.json(),
                headers: (key) => request.headers.get(key),
                method: () => request.method,
                url: () => {
                    // Attempt to get host separately as Astro will warn if
                    // not in server output mode
                    const host = request.headers.get("host");
                    if (!host || host.length === 0) {
                        throw new Error(`Could not access Astro.request.headers.host. Please change your Astro config to use "server" or "hybrid" output mode.`);
                    }
                    return new URL(request.url, `https://${host || ""}`);
                },
                transformResponse: ({ body, status, headers }) => {
                    return new Response(body, { status, headers });
                },
            };
        } }));
    const requestHandler = commHandler.createHandler();
    return Object.defineProperties(requestHandler, {
        GET: { value: requestHandler },
        POST: { value: requestHandler },
        PUT: { value: requestHandler },
    });
};
exports.serve = serve;
