// inngest/digitalocean.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.serve = exports.frameworkName = void 0;
const InngestCommHandler_1 = require("./components/InngestCommHandler");
exports.frameworkName = "digitalocean";
const serve = (options) => {
    const handler = new InngestCommHandler_1.InngestCommHandler(Object.assign(Object.assign({ frameworkName: exports.frameworkName }, options), { handler: (main = {}) => {
            const { http = { method: "GET", headers: {}, path: "" } } = main, data = __rest(main, ["http"]);
            return {
                body: () => data || {},
                headers: (key) => { var _a; return (_a = http === null || http === void 0 ? void 0 : http.headers) === null || _a === void 0 ? void 0 : _a[key]; },
                method: () => http.method,
                url: () => new URL(`${options.serveHost}${options.servePath || "/"}`),
                queryString: (key) => main[key],
                transformResponse: (res) => res,
            };
        } }));
    return handler.createHandler();
};
exports.serve = serve;
